wnfs/src/private/forest.rs

```diff
@@ -298,22 +288,205 @@ impl PrivateForest {
     }
 }
 
+impl<H> Hamt<Namefilter, BTreeSet<Cid>, H>
+where
+    H: Hasher + fmt::Debug + Clone + 'static,
+{
+    /// Merges a private forest with another. If there is a conflict with values, it combines the two values in the final merge node
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::rc::Rc;
+    /// use chrono::{Utc, Days};
+    /// use rand::{thread_rng, Rng};
+    /// use wnfs::{
+    ///     private::{PrivateForest, PrivateRef}, PrivateNode,
+    ///     BlockStore, MemoryBlockStore, Namefilter, PrivateDirectory, PrivateOpResult,
+    /// };
+    ///
+    /// #[async_std::main]
+    /// async fn main() {
+    ///     let store = &mut MemoryBlockStore::default();
+    ///     let rng = &mut thread_rng();
+    ///
+    ///     let ratchet_seed = rng.gen::<[u8; 32]>();
+    ///     let inumber = rng.gen::<[u8; 32]>();
+    ///
+    ///     let main_forest = Rc::new(PrivateForest::new());
+    ///     let root_dir = Rc::new(PrivateDirectory::with_seed(
+    ///         Namefilter::default(),
+    ///         Utc::now(),
+    ///         ratchet_seed,
+    ///         inumber
+    ///     ));
+    ///     let main_forest = main_forest
+    ///         .put(
+    ///             root_dir.header.get_saturated_name(),
+    ///             &root_dir.header.get_private_ref(),
+    ///             &PrivateNode::Dir(Rc::clone(&root_dir)),
+    ///             store,
+    ///             rng
+    ///         )
+    ///         .await
+    ///         .unwrap();
+    ///
+    ///     let other_forest = Rc::new(PrivateForest::new());
+    ///     let root_dir = Rc::new(PrivateDirectory::with_seed(
+    ///         Namefilter::default(),
+    ///         Utc::now().checked_add_days(Days::new(1)).unwrap(),
+    ///         ratchet_seed,
+    ///         inumber
+    ///     ));
+    ///     let other_forest = other_forest
+    ///         .put(
+    ///             root_dir.header.get_saturated_name(),
+    ///             &root_dir.header.get_private_ref(),
+    ///             &PrivateNode::Dir(Rc::clone(&root_dir)),
+    ///             store,
+    ///             rng
+    ///         )
+    ///         .await
+    ///         .unwrap();
+    ///
+    ///     let merge_forest = main_forest.merge(&other_forest, store).await.unwrap();
+    ///
+    ///     assert_eq!(
+    ///         2,
+    ///         merge_forest
+    ///             .root
+    ///             .get(&root_dir.header.get_saturated_name(), store)
+    ///             .await
+    ///             .unwrap()
+    ///             .unwrap()
+    ///             .len()
+    ///     );
+    /// }
+    /// ```
+    pub async fn merge<B: BlockStore>(&self, other: &Self, store: &mut B) -> Result<Self> {
```

Comment:
Ooooh.

For a moment I was confused by the type signature here. "Why doesn't in specify how to merge value pairs?".
This is within an implementation on the more specific type `Hamt<Namefilter, BTreeSet<Cid>>`.

Not bad for a start, but perhaps it would make sense to turn this into a higher-order function and take something like `fn(V, V) -> V` as a parameter. (That function then *has* to be associative, commutative and idempotent as well. collecting after `BTreeSet::union` has that property.)

Reply:
I agree. Thought about that.